src/core/writer.js

```diff
@@ -150,52 +150,68 @@ function writeXFADataForAcroform(str, newRefs) {
   return buffer.join("");
 }
 
-function updateXFA({
-  xfaData,
-  xfaDatasetsRef,
-  hasXfaDatasetsEntry,
-  acroFormRef,
+function updateAcroform({
+  xref,
   acroForm,
+  acroFormRef,
+  hasXfa,
+  hasXfaDatasetsEntry,
+  xfaDatasetsRef,
+  needAppearances,
   newRefs,
-  xref,
-  xrefInfo,
 }) {
-  if (xref === null) {
+  if (hasXfa && !hasXfaDatasetsEntry && !xfaDatasetsRef) {
+    warn("XFA - Cannot save it");
+  }
+
+  if (!needAppearances && (!hasXfa || !xfaDatasetsRef)) {
     return;
   }
 
-  if (!hasXfaDatasetsEntry) {
-    if (!acroFormRef) {
-      warn("XFA - Cannot save it");
-      return;
-    }
+  // Clone the acroForm.
+  const dict = new Dict(xref);
+  for (const key of acroForm.getKeys()) {
+    dict.set(key, acroForm.getRaw(key));
+  }
 
+  if (hasXfa && !hasXfaDatasetsEntry) {
     // We've a XFA array which doesn't contain a datasets entry.
     // So we'll update the AcroForm dictionary to have an XFA containing
     // the datasets.
-    const oldXfa = acroForm.get("XFA");
-    const newXfa = oldXfa.slice();
+    const newXfa = acroForm.get("XFA").slice();
     newXfa.splice(2, 0, "datasets");
     newXfa.splice(3, 0, xfaDatasetsRef);
 
-    acroForm.set("XFA", newXfa);
+    dict.set("XFA", newXfa);
+  }
 
-    const encrypt = xref.encrypt;
-    let transform = null;
-    if (encrypt) {
-      transform = encrypt.createCipherTransform(
-        acroFormRef.num,
-        acroFormRef.gen
-      );
-    }
+  if (needAppearances) {
+    dict.set("NeedAppearances", true);
+  }
 
-    const buffer = [`${acroFormRef.num} ${acroFormRef.gen} obj\n`];
-    writeDict(acroForm, buffer, transform);
-    buffer.push("\n");
+  const encrypt = xref.encrypt;
+  let transform = null;
+  if (encrypt) {
+    transform = encrypt.createCipherTransform(acroFormRef.num, acroFormRef.gen);
+  }
+
+  const buffer = [];
+  writeObject(acroFormRef, dict, buffer, transform);
 
-    acroForm.set("XFA", oldXfa);
+  newRefs.push({ ref: acroFormRef, data: buffer.join("") });
+}
 
-    newRefs.push({ ref: acroFormRef, data: buffer.join("") });
+function updateXFA({
+  xfaData,
+  xfaDatasetsRef,
+  hasXfaDatasetsEntry,
+  acroFormRef,
+  acroForm,
+  newRefs,
+  xref,
+}) {
+  if (xref === null) {
+    return;
```

Comment:
I know that you're only moving existing code around, but can this actually happen in practice?